### PR TITLE
 cdc: log: fill cdc$deleted_ columns in pre-images 

### PR DESCRIPTION
**Pull request by avelanarius**
Before this change, `cdc$deleted_` columns were all `NULL` in pre-images. Lack of such information made it hard to correctly interpret the pre-image rows, for example:

```
INSERT INTO tbl(pk, ck, v, v2) VALUES (1, 1, null, 1);
INSERT INTO tbl(pk, ck, v2) VALUES (1, 1, 1);
```

For this example, pre-image generated for the second operation would look like this (in both `true` and `full` pre-image mode):

```
pk=1, ck=1, v=NULL, cdc$deleted_v=NULL, v2=1
```

`v=NULL` has two meanings:
1. If pre-image was in `true` mode, `v=NULL` describes that v was not affected (affected columns: pk, ck, v2).
2. If pre-image was in `full` mode, `v=NULL` describes that v was equal to `NULL` in the pre-image.

Therefore, to properly decode pre-images you would need to know in which mode pre-image was configured on the CDC-enabled table at the moment this CDC log row was inserted. There is no way to determine such information (you can only check a current mode of pre-image).

A solution to this problem is to fill in the `cdc$deleted_` columns for pre-images. After this PR, for the `INSERT` described above, CDC now generates the following log row:

If in pre-image 'true' mode:
```
pk=1, ck=1, v=NULL, cdc$deleted_v=NULL, v2=1
```

If in pre-image 'full' mode:
```
pk=1, ck=1, v=NULL, cdc$deleted_v=true, v2=1
```

A client library now can properly decode a pre-image row. If it sees a `NULL` value, it can now check the `cdc$deleted_` column to determine if this `NULL` value was a part of pre-image or it was omitted due to not being an affected column in the delta operation.

No such change is necessary for the post-image rows, as those images are always generated in the `full` mode.

### Additional example
Additional example of trouble decoding pre-images before this change.
tbl2 - `true` pre-image mode, tbl3 - `full` pre-image mode:

```
INSERT INTO tbl2(pk, ck, v, v2) VALUES (1, 1, 5, 1);
INSERT INTO tbl3(pk, ck, v, v2) VALUES (1, 1, null, 1);
```

```
INSERT INTO tbl2(pk, ck, v2) VALUES (1, 1, 1);
```
generated pre-image:
```
pk=1, ck=1, v=NULL, cdc$deleted_v=NULL, v2=1
```

```
INSERT INTO tbl3(pk, ck, v2) VALUES (1, 1, 1);
```

generated pre-image:
```
pk=1, ck=1, v=NULL, cdc$deleted_v=NULL, v2=1
```

Both pre-images look the same, but:
1. `v=NULL` in tbl2 describes v being omitted from the pre-image.
2. `v=NULL` in tbl3 described v being `NULL` in the pre-image.

### CODE REVIEW

**Review by nyh — APPROVED** [review]
Looks good to me, BUT because this involves an API change, I'll appreciate also a review from one of the other CDC developers that this is an API change in the right direction.
If you consider this to be a necessary fix, not some "nice to have" improvement, please also create an issue for this and refer to it, so it can be backported.
Also, please make sure that you update https://docs.scylladb.com/using-scylla/cdc/cdc-preimages/ for the new change - because it explicitly documents the old behavior and must be updated. As usual, it would have been better if the document was inside the repository and this fix and the doc fix could go together, but we can't.

**Comment by nyh** [comment]
By the way, since the Alternator Streams **uses** the CDC API, any changes to that API might affect the Alternator Streams implementation, and we need to verify that it doesn't break anything. I confirmed that Alternator Streams tests work after this patch, but I don't know if the tests we have are sensitive enough in this area of a preimage with unset columns (@elcallio?). 

I **hope** that Alternator won't have a problem with this change, because as you said the vague meaning of "v=NULL" in a preimage happens when you don't know whether the preimage was full or true (and in Alternator, we always know it's full). But, we need to check that the extra column `deleted$v` doesn't cause problems (e.g., it mustn't be copied to the preimage item that Alternator generates! I don't know if we already exclude all those "$" columns).

Another note about Alternator: In Alternator we usually have map elements instead of full-fledged columns, so all this difference between "full" and "true" preimage is subtle (if there's any change to the map, the entire map appears anyway). If we have a problem here, it might be visible only in the case of GSI - where one of the non-key columns (the one who is the GSI's key) becomes a real Scylla column and may have this `deleted$v` thing.

**Comment by nyh** [comment]
> Another note about Alternator: In Alternator we usually have map elements instead of full-fledged columns, so all this difference between "full" and "true" preimage is subtle (if there's any change to the map, the entire map appears anyway). If we have a problem here, it might be visible only in the case of LSI - where one of the non-key columns (the one who is the GSI's key) becomes a real Scylla column and may have this `deleted$v` thing.

I'll write myself a test for this edge case for Alternator, and report back my finding if anything breaks.

**Comment by nyh** [comment]
I verified that Alternator Streams is ok with the change to the API:

1. First a code review: The code ignores the new `delete$v` column because in `executor::get_records()` we only output columns originally in the base table, and ignore any other columns which CDC adds.
2. A test: the following test that exercises this case (preimage with null value) and passes - we don't get any unwanted "delete$k" column added to the Alternator Streams' GetRecords ouput:

```python
# This test is the same as the previous (test_streams_updateitem_old_image_lsi)
# except that the *old* value of the LSI key k is missing. If CDC adds a
# special deleted$k marker for a missing column in the preimage (see PR
# #8568), we want to check that Alternator Streams remembers to remove this
# marker in its output.
def test_streams_updateitem_old_image_lsi_missing_column(test_table_ss_old_image_and_lsi, dynamodbstreams):
    def do_updates(table, p, c):
        events = []
        # Note that we do *not* set the "k" attribute (the LSI key)
        table.update_item(Key={'p': p, 'c': c},
            UpdateExpression='SET x = :x',
            ExpressionAttributeValues={':x': 2})
        events.append(['*', {'p': p, 'c': c}, None, {'p': p, 'c': c, 'x': 2}])
        table.update_item(Key={'p': p, 'c': c},
            UpdateExpression='SET y = :y', ExpressionAttributeValues={':y': 3})
        # "k" should be missing from OldImage, because it wasn't set earlier.
        events.append(['MODIFY', {'p': p, 'c': c}, {'p': p, 'c': c, 'x': 2}, {'p': p, 'c': c, 'x': 2, 'y': 3}])
        return events
    do_test(test_table_ss_old_image_and_lsi, dynamodbstreams, do_updates, 'OLD_IMAGE')
```

File: cdc/log.cc [hunk 1]
```diff
@@ -1518,6 +1518,18 @@ class transformer final : public change_processor {
         auto process_cell = [&, this] (const column_definition& cdef) {
             if (auto current = get_col_from_row_state(row_state, cdef)) {
                 _builder->set_value(image_ck, cdef, *current);
+            } else if (op == operation::pre_image) {
```
Comment (haaawk) [inline cdc/log.cc:1524]: I would extract the `if` outside the lambda to check it once not for every column. So auto process_cell = op == operation::pre_image ? ... : ...;

**Review by haaawk — APPROVED** [review]
LGTM. I think the change in the API makes a lot of sense. Thanks @nyh for checking the Alternator Streams compatibility.

**Comment by nyh** [comment]
> LGTM. I think the change in the API makes a lot of sense. 

Ok, good. I have to be honest that the more I looked into the Alternator Streams as a user of this API, the more I started to doubt if this API change is important. I suggest that we create an issue to explain why or by whom this change was requested.

The thing is, I think that a user - like in my case the Alternator Streams code - knows what he asked for. Alternator Streams knows that it asked for preimage=full, so it knew exactly what x=null in a preimage meant - that x is really missing in the preimage. The extra `deleted$x` you add now is a (tiny) waste of space, and rather than make use of it - Alternator Streams just needed to make sure now that we ignore it (and the good news is that we already did, so no change was necessary).

> Thanks @nyh for checking the Alternator Streams compatibility.

You're welcome :-)
Please do consider creating an issue to go along with this change, and update https://docs.scylladb.com/using-scylla/cdc/cdc-preimages/ to explain the new behavior instead of the old one.

**Comment by haaawk** [comment]
> The thing is, I think that a user - like in my case the Alternator Streams code - knows what he asked for. Alternator Streams knows that it asked for preimage=full, so it knew exactly what x=null in a preimage meant - that x is really missing in the preimage. The extra `deleted$x` you add now is a (tiny) waste of space, and rather than make use of it - Alternator Streams just needed to make sure now that we ignore it (and the good news is that we already did, so no change was necessary).

We have a generic consumers like our library or a Kafka connector and they don't know what was set by the user but still have to work. In theory they could check the schema but the schema can change and then we're in this complicated consistency related situation. It's much easier to just make the CDC Log entries self explainable.

File: cdc/log.cc (hunk 1 above)
Comment (avelanarius) [reply cdc/log.cc:1524]: I don't see a clean way to do it, without code duplication.

**Comment by avelanarius** [comment]
v2: In the amended commits I made two minor changes:
1. Added missing newline at the end of `cdc_test.cc`
2. Clarified the comment in the test description. "iff a read actually happened" before was not entirely accurate.

    before: "Pre-image rows should set cdc$deleted_ column for columns with NULL values iff a read actually happened ('full' preimage or column was affected)."

    after: "Pre-image rows should set cdc$deleted_ column: 1. If pre-image is in 'full' mode, for all NULL columns in the read pre-image. 2. If pre-image is in 'true' mode, for all columns that are both affected and have NULL value in the read pre-image."

@nyh Should I also add your Alternator Streams test to my PR? (of course with mention of your authorship in the commit message)

**Review by piodul — APPROVED** [review]
LGTM

**Comment by nyh** [comment]
> @nyh Should I also add your Alternator Streams test to my PR? (of course with mention of your authorship in the commit message)

I'll improve the test comments a bit, and send such a patch myself. Thanks.

**Comment by nyh** [comment]
I merged this PR, and sent a patch to the mailing list titled "test/alternator: test for change in CDC preimage" with the new Alternator test (which passes).

@avelanarius / @haaawk please do plan to send a patch to the documentation https://docs.scylladb.com/using-scylla/cdc/cdc-preimages/ as soon as possible, because it explicitly mentions the old behavior (claiming there is no "deleted" in preimage) - and if you don't edit now, you'll never remember to do it later.

**Comment by avelanarius** [comment]
This a patch for the documentation: https://github.com/scylladb/scylla-docs/pull/348

**Comment by nyh** [comment]
On Tue, May 4, 2021 at 3:51 PM Piotr Grabowski ***@***.***>
wrote:

> This a patch for the documentation: scylladb/scylla-docs#348
> <https://github.com/scylladb/scylla-docs/issues/348>
>

Wrong link?

**Comment by avelanarius** [comment]
> On Tue, May 4, 2021 at 3:51 PM Piotr Grabowski ***@***.***> wrote: This a patch for the documentation: [scylladb/scylla-docs#348](https://github.com/scylladb/scylla-docs/issues/348) <[scylladb/scylla-docs#348](https://github.com/scylladb/scylla-docs/issues/348)>
> Wrong link?

Yes, the last character got cut-off: https://github.com/scylladb/scylla-docs/pull/3480

**Pull request closed without merging**